clients/kafka-common/src/main/java/org/eclipse/hono/kafka/client/consumer/AbstractAtLeastOnceKafkaConsumer.java

```diff
@@ -0,0 +1,231 @@
+ * Already processed messages will then be again in the next batch and be passed to the message handler. If
+ * de-duplication of messages is required, it must be handled by the message handler.
+ * <p>
+ * The consumer starts consuming when {@link #start()} is invoked. It needs to be closed by invoking {@link #stop()} to
+ * release the resources.
+ * <p>
+ * </p>
+ *
+ * ERROR CASES:
+ * <p>
+ * Errors can happen when polling, in message processing, and when committing the offset to Kafka.
+ *
+ * If a fatal error occurs, the underlying Kafka consumer will be closed and the close-handler invoked with an exception
+ * indicating the cause.
+ * <p>
+ * If {@link KafkaConsumer#poll(Duration, Handler)} fails during the start, {@link #start()} will return a failed
+ * future. For subsequent {@code poll} operations, the Kafka consumer will be closed and the close handler will be
+ * passed a {@link KafkaConsumerPollException}.
+ *
+ * <p>
+ * If the message processing fails because either {@link #createMessage(KafkaConsumerRecord)} or the message handler
+ * throws an exception, the Kafka consumer will be closed and the exception will be passed to the close handler.
+ * <p>
+ * If {@link KafkaConsumer#commit(Handler)} times out, the commit will be retried once. If the retry fails or the commit
+ * fails with another exception, the Kafka consumer will be closed and the close handler will be passed a
+ * {@link KafkaConsumerCommitException}. For example, commits could regularly fail during consumer rebalance.
+ *
+ * @param <T> The type of message to be created from the Kafka records.
+ */
+public abstract class AbstractAtLeastOnceKafkaConsumer<T> implements Lifecycle {
+
+    // TODO add unit test
+
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractAtLeastOnceKafkaConsumer.class);
+
+    private final KafkaConsumer<String, Buffer> kafkaConsumer;
+    private final String topic;
+    private final Consumer<T> messageHandler;
+    private final Handler<Throwable> closeHandler;
+    private final Duration pollTimeout;
+
+    /**
+     * Creates a new consumer.
+     *
+     * @param kafkaConsumer The Kafka consumer to be used to consume records
+     * @param topic The Kafka topic to consume records from.
+     * @param messageHandler The function to be invoked for each message created from a record.
+     * @param closeHandler The handler to be invoked when the Kafka consumer has been closed due to an error.
+     * @param pollTimeout The maximal number of milliseconds to wait for messages during a poll operation.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     * @see #start()
+     * @see #stop()
+     */
+    public AbstractAtLeastOnceKafkaConsumer(final KafkaConsumer<String, Buffer> kafkaConsumer, final String topic,
+            final Consumer<T> messageHandler, final Handler<Throwable> closeHandler, final long pollTimeout) {
```

Comment:
The way the `AbstractAtLeastOnceKafkaConsumer` currently is implemented means the given `kafkaConsumer` can only be used exclusively with this `AbstractAtLeastOnceKafkaConsumer` instance, and the one given topic, right?

I think that should be pointed out here. And for scenarios where multiple topics are needed, corresponding constructors with `Set<String> topics` and `Pattern topicPattern` parameters would be good.

Reply:
> The way the `AbstractAtLeastOnceKafkaConsumer` currently is implemented means the given `kafkaConsumer` can only be used exclusively with this `AbstractAtLeastOnceKafkaConsumer` instance, and the one given topic, right?

Yes. Providing the native `KafkaConsumer` as a parameter is mainly for testability (in unit tests I would use `org.apache.kafka.clients.consumer.MockConsumer` here). Maybe it would be an improvement to add a second constructor without the parameter. Then the class would manage the `KafkaConsumer` while I can still provide the MockConsumer in unit tests. WDYT?